### PR TITLE
feat: optimize read_rows

### PR DESCRIPTION
**Pull request by daniel-sanche**
Rewrite the row merger logic to improve throughput

This change brings us to ~75k RPS with retries, ~90 with retries disabled (retry logic will be optimized in a future PR)

Note: the idle timeout has been removed as a part of these changes. Other features are preserved

https://github.com/googleapis/python-bigtable/issues/857

### CODE REVIEW

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 1]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:59]: Do you need to add slots? I think _last_yielded_row_key will accessed fairly often

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 2]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
-        self,
-        gapic_fn: Callable[..., Awaitable[AsyncIterable[ReadRowsResponse]]],
-        timeout_generator: Iterator[float],
-        total_row_limit: int,
-    ) -> AsyncGenerator[Row, None]:
+        def on_error_fn(exc):
+            if self._predicate(exc):
+                transient_errors.append(exc)
+
+        retry_fn = AsyncRetryableGenerator(
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:102]: nit. this isnt a function. Maybe retryable_gen?

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 3]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
+            self.operation_timeout,
+            on_error_fn,
+        )
+        try:
+            async for row in retry_fn:
+                yield row
+                if self._remaining_count is not None:
+                    self._remaining_count -= 1
+                    if self._remaining_count < 0:
+                        raise RuntimeError("emit count exceeds row limit")
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:115]: Should this be a more specific error? Maybe add a todo to improve it in the future

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 4]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
+        except core_exceptions.RetryError:
+            self._raise_retry_error(transient_errors)
+        except GeneratorExit:
+            # propagate close to wrapped generator
+            await retry_fn.aclose()
+
+    def read_rows_attempt(self) -> AsyncGenerator[Row, None]:
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:122]: should this be prefixed with _?

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 5]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
-        return self
-
-    async def __anext__(self) -> Row:
-        """Implements the AsyncIterator interface"""
-        if self._stream is not None:
-            return await self._stream.__anext__()
-        else:
-            raise asyncio.InvalidStateError("stream is closed")
+        self._last_yielded_row_key: bytes | None = None
+        self._remaining_count = self.request.rows_limit or None
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:90]: should this have a type to be consistent?

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 6]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
+                                raise InvalidChunk("row key changed mid cell")
+
+                            if c.reset_row:
+                                raise _ResetRow(c)
+                            buffer.append(c.value)
+                        value = b"".join(buffer)
+                    if family is None:
+                        raise InvalidChunk("missing family")
+                    if qualifier is None:
+                        raise InvalidChunk("missing qualifier")
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:281]: Can you move this up near the other family & qualifier checks? L246

File: google/cloud/bigtable/data/_async/_read_rows.py [hunk 7]
```diff
@@ -59,158 +59,253 @@ class _ReadRowsOperationAsync(AsyncIterable[Row]):
...
-                    self._emit_count += 1
-                self._last_emitted_row_key = new_item.row_key
-                if total_row_limit and self._emit_count >= total_row_limit:
-                    return
-        except (Exception, GeneratorExit) as exc:
-            # ensure grpc stream is closed
-            new_gapic_stream.cancel()
-            raise exc
+                    raise InvalidChunk("reset row with data")
+                continue
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/_async/_read_rows.py:300]: I think you have to reset the shared family & qualifier variables

File: google/cloud/bigtable/data/__init__.py [hunk 8]
```diff
@@ -66,7 +67,7 @@
     "DeleteAllFromRow",
     "Row",
     "Cell",
-    "ReadRowsAsyncIterator",
+    # "ReadRowsAsyncIterator",
```
Comment (igorbernstein2) [inline google/cloud/bigtable/data/__init__.py:70]: should this be removed? along with the other commented out lines

File: tests/system/data/test_system.py [hunk 9]
```diff
@@ -803,46 +749,51 @@ async def test_read_rows_with_filter(table, temp_rows):
...
-        assert "inactivity" in str(e)
-        assert "idle_timeout=0.1" in str(e)
+#     generator = await table.read_rows_stream({})
+#     await generator._start_idle_timer(0.05)
+#     await asyncio.sleep(0.2)
+#     assert generator.active is False
+#     with pytest.raises(IdleTimeout) as e:
+#         await generator.__anext__()
+#         assert "inactivity" in str(e)
+#         assert "idle_timeout=0.1" in str(e)
```
Comment (igorbernstein2) [inline tests/system/data/test_system.py:793]: please remove

File: tests/unit/data/_async/test__read_rows.py [hunk 10]
```diff
@@ -429,197 +363,197 @@ async def aclose(self):
...
-        assert iterator.active is False
-        assert isinstance(iterator._error, StopAsyncIteration)
-        assert iterator._idle_timeout_task is None
-        with pytest.raises(StopAsyncIteration) as e:
-            await iterator.__anext__()
-            assert "closed" in str(e.value)
-        # timeout task should be cancelled
-        with pytest.raises(asyncio.CancelledError):
-            await timeout_task
+# class TestReadRowsAsyncIterator:
```
Comment (igorbernstein2) [inline tests/unit/data/_async/test__read_rows.py:366]: ?

File: tests/unit/data/_async/test_client.py [hunk 11]
```diff
@@ -1351,54 +1338,54 @@ async def test_read_rows_attempt_timeout(
...
-            with pytest.raises(IdleTimeout) as e:
-                await gen.__anext__()
-
-            expected_msg = (
-                "Timed out waiting for next Row to be consumed. (idle_timeout=0.1s)"
-            )
-            assert e.value.message == expected_msg
-            aclose.assert_called_once()
-            aclose.assert_awaited()
+    # @pytest.mark.asyncio
```
Comment (igorbernstein2) [inline tests/unit/data/_async/test_client.py:1341]: please remove

File: google/cloud/bigtable/data/_async/_read_rows.py (hunk 1 above)
Comment (daniel-sanche) [reply google/cloud/bigtable/data/_async/_read_rows.py:59]: Good point, fixed. I didn't see a noticeable a perf bump in my run after adding slots, but still seems like a good change

File: google/cloud/bigtable/data/_async/_read_rows.py (hunk 7 above)
Comment (daniel-sanche) [reply google/cloud/bigtable/data/_async/_read_rows.py:311]: Those should be reset on each row iteration on L232 and L233

File: google/cloud/bigtable/data/_async/_read_rows.py (hunk 3 above)
Comment (daniel-sanche) [reply google/cloud/bigtable/data/_async/_read_rows.py:126]: Sure, I opened an issue to track this here: https://github.com/googleapis/python-bigtable/issues/853

**Pull request merged**